qsv: improved full path via vpp

galinart opened the pull request:
* Integrated latest oneVPL v2023.3.1 and latest VPP patches from FFmpeg master branch.
* Allow more configurations that did not work properly via full path be processed via VPP
* [UPD] Refactored cropscale filter and moved format conversions to format filter  
* [UPD] tonemap filter will be added as separate PR
* [UPD] fixed black buffer creation for vpp pipelines 

[Intel Arc driver](https://www.intel.com/content/www/us/en/download/726609/intel-arc-iris-xe-graphics-whql-windows.html?
) .4577 is recommended

galinart committed:
qsv: enable optional full path via system memory

Co-authored-by: Vladyslav Sosunovych <vladyslav.sosunovych@intel.com>

galinart committed:
qsv: add ffmpeg latest vpp patches

Co-authored-by: Vladyslav Sosunovych <vladyslav.sosunovych@intel.com>

galinart committed:
qsv: enable more conversions via hw vpp in case of full path

Co-authored-by: Vladyslav Sosunovych <vladyslav.sosunovych@intel.com>

galinart committed:
qsv: update oneVPL to version v2023.3.1

Co-authored-by: Vladyslav Sosunovych <vladyslav.sosunovych@intel.com>

galinart committed:
qsv: adjust hyper encode default settings

Co-authored-by: Vladyslav Sosunovych <vladyslav.sosunovych@intel.com>

galad87 commented:
Could the tonemap and range/pixel format conversion be moved to their respective filters? Cramming everything in cropscale or adding extra checks in work.c and sync.c doesn't look too clean.
Then I guess they can be converted back to a single VPP filter, but it would be better to improve the existing format.c and tonemap.c filters, it's hard to reason about things with all these special QSV options that works differently than everything else.

galinart commented:
@galad87 refactored a bit the code. Thanks for the feedback. I will look how to move color conversion and tonemap into corresponding filters and come back with another PR. For now, may be it worth to split the PR into two.

galad87 commented:
Are all these patches in FFmpeg master branch? If so we could just update to the latest FFmpeg git for now, until 6.1 is out (it should be sooner than later but I guess the drama got in the way).

galinart commented on contrib/libvpl/module.defs:
```diff
@@ -1,11 +1,11 @@
-LIBVPL.FETCH.basename  = oneVPL-2023.1.0.tar.gz
-LIBVPL.EXTRACT.tarbase = oneVPL-2023.1.0
+LIBVPL.FETCH.url       = https://github.com/oneapi-src/oneVPL/archive/refs/tags/v2023.3.1.tar.gz
+LIBVPL.FETCH.url      += https://github.com/HandBrake/HandBrake-contribs/releases/download/contribs/oneVPL-2023.3.1.tar.gz
```
TODO: to be added to HandBrake contrib

galinart commented:
Not sure that I have write permissions to HandBrake contrib repo to upload FFmpeg git archive. Alternatively we can use HEAD snapshot from https://ffmpeg.org/releases/ffmpeg-snapshot.tar.bz2

BTW I have updated PR and split the logic into 2 filters. Tonemap filter will be added as separate filter in other PR later.

galad87 commented on libhb/work.c:
```diff
@@ -1477,6 +1468,15 @@ static void sanitize_filter_list_post(hb_job_t *job)
         free(settings);
     }
+
+#if HB_PROJECT_FEATURE_QSV && (defined( _WIN32 ) || defined( __MINGW32__ ))
```
Is this needed? At this point it should be already known since the hb_get_best_hw_pix_fmt() if hw filters can be used or not.

galinart replied:
yes, this is needed since we added format filter and modify pipeline in post stage e.g. 8b -> 10b. This does not impact the hw_pix_fmt decision here and format filter is hw supported but important to know if hw filters are actually used or not. Also there is scenario where we don't have format filter e.g. 8b ->8b but have sw filter in the middle. In this case pre stage sanitize is needed for correct hw_pix_fmt.

galad87 replied:
I see, I wonder if most of those hb_qsv_hw_filters_are_enabled() can be replaced with hw_pix_fmt == AV_PIX_FMT_QSV.

galinart replied:
In some places that is true, but AV_PIX_FMT_QSV does not guarantee that HW filters are used or not. That only guarantee that full path via video memory is used. 

I changed the logic a bit and removed one of sanitize_qsv call from work.c

galinart replied:
Reverted some of the changes since HB GUI does scan and generate previews which is different from CLI behaviour.

galad87 commented on libhb/work.c:
```diff
@@ -1703,6 +1703,12 @@ static void do_job(hb_job_t *job)
+#if HB_PROJECT_FEATURE_QSV
+        if (hb_qsv_full_path_is_enabled(job))
+        {
+            init.color_range = (job->qsv.ctx->out_range == AVCOL_RANGE_UNSPECIFIED) ? title->color_range : job->qsv.ctx->out_range;
```
There isn't a way to general way manually set color_range yet, and it's always set to limited unless encoding Dolby Vision profile 5 (which isn't supported in QSV yet) so I don't know about this. Maybe it could be left out until we have a general way to set it. I was thinking of adding an additional popup menu in the Video panel to make it selectable.

galinart replied:
For now in case if the input is full range and sw pipeline is limited, don't we need to do a explicit conversion from full -> limited range all the time? I don't see that sw filter inserted for this automatically and looks that full range input processed as limited. For QSV I just wanted to allow use input true color range as is if we don't use sw filters anyway.

galad87 replied:
The conversion is done in decavcodec.c right now for the sw path, together with size and rotation.

galad87 replied:
QSV full path doesn't do any conversion at the moment, it's related to issue #5179.
Anyway, no harm in keeping those few lines for now I guess.

galinart replied:
> The conversion is done in decavcodec.c right now for the sw path, together with size and rotation.

Why is conversion needed for e.g. 10b full range sw decoder -> (no filters) -> 10b full range sw encoder pipelines?

We are looking at #5179 separately.

galad87 replied:
It's not, decavcodec.c converts only If needed, it's just that's HandBrake always set the job range to limited (expect for Dolby Vision 5), and there is no UI to set the preferred range apart from your specific QSV hack above.

galad87 reviewed (commented):
Thanks, I added two minor feedbacks. I don't know if I will have time to test it in the few next days.
About the FFmpeg upgrade, I guess it will better to get this in and then try to upgrade it.

galinart committed:
qsv: sanitize filters in pre and post stages

galinart committed:
qsv: simplify logic for hw filters usage

galinart commented:
Can you please upload FFmpeg and oneVPL archives to contrib when you have a time?

galad87 commented:
Uploaded ffmpeg-2023-8-17.tar.bz2 and oneVPL-2023.3.1.tar.gz

galinart commented:
A lot of patches(not VPP) from contrib conflicts with snapshot and not part of the master and requires manual patch modifications. I would recommend to do this as separate PR to upgrade to latest FFmpeg with extensive testing from other folks as well.

galad87 commented:
Yup, I will probably have time next week for the FFmpeg update.
I'll do a quick test of this PR tomorrow.

galad87 commented:
Tested it and the vpp format conversion seems to work.
I got this log entries, probably from the generated black buffers from the usual bbb_sunflower_1080p_30fps_normal.mp4:

> [Parsed_vpp_qsv_0 @ 0000021ca8fd0000] QSVVPP gets a wrong frame.
> [Parsed_vpp_qsv_0 @ 0000021ca8fd0000] Failed to submit frame on input[0]
> [Parsed_vpp_qsv_0 @ 0000021ca8fd0000] QSVVPP gets a wrong frame.
> [Parsed_vpp_qsv_0 @ 0000021ca8fd0000] Failed to submit frame on input[0]

Here's the complete activity log: https://gist.github.com/galad87/e50d771a50fd7fb001e00e068377e993

But the initial black frames seem ok in the output.

galinart committed:
qsv: fix creation of black buffers

galinart commented:
It was in master and I have never seen this until changed the verbosity level. Fixed now.

galad87 commented:
@sr55 do you want to check this too before merging?

sr55 commented:
Sure, I'll try make time for that tomorrow.

sr55 commented:
Ran some quick checks and it looks fine to me.
